### PR TITLE
Kill startLine & endLine mutants

### PR DESCRIPTION
**Pull request by staabm**
lets see how https://github.com/nikic/PHP-Parser/pull/985 affects mutations

refs https://github.com/Roave/BetterReflection/issues/1382

### CODE REVIEW

**Comment by staabm** [comment]
@kukulich it looks like it doesn't improve mutations. I guess phpdoc-types are not useful for that (and only native types are taken into account?)

**Comment by Ocramius** [comment]
Phpdoc types **are** considered: need to likely checked in more depth 🤔

Thanks for trying though!

**Comment by kukulich** [comment]
@staabm 

I think this can work:

```
$startLine = $node->getStartLine()
if ($startLine !== -1) {
    $this->startLine = $startLine;
}
```

No `assert` -> no mutation and no mutant :) Psalm/Phpstan should be happy too thanks to your change in PhpParser.

The current solution probably does not work because `if ($node->hasAttribute('startLine')) {` does not tell Psalm that `$node->getStartLine()` is only positive-int. That's why the `assert` is not reported by Psalm as useless.

**Comment by staabm** [comment]
Thank you. I will give your idea a try tomorrow 

(Btw: the php-parser PR was merged)

**Comment by staabm** [comment]
the psalm error is a bug, reported in https://github.com/vimeo/psalm/issues/10826

**Comment by kukulich** [comment]
It looks it's very complicated to kill one mutant 😄

**Comment by staabm** [comment]
> It looks it's very complicated to kill one mutant 😄

hehe - agree, its a lot of change. but I feel the code is easier to read after the PR.

**Comment by kukulich** [comment]
> I feel the code is easier to read after the PR.

I agree :)

**Comment by staabm** [comment]
@Ocramius mutants will be killed on the next php-parser release. I think its good to go.

File: psalm-baseline.xml [hunk 1]
```diff
@@ -149,16 +149,12 @@
...
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
-      <code><![CDATA[hasAttribute]]></code>
       <code><![CDATA[traverse]]></code>
     </ImpureMethodCall>
+    <PropertyTypeCoercion>
```
Comment (staabm) [inline psalm-baseline.xml:154]: these baselined errors are related to https://github.com/Roave/BetterReflection/pull/1407#issuecomment-1996573163

**Review by Ocramius — APPROVED** [review]
```diff
- 4164 mutations were generated:
+ 4158 mutations were generated:
-    3333 mutants were killed
+    3329 mutants were killed
       0 mutants were configured to be ignored
      14 mutants were not covered by tests
-      10 covered mutants were not detected
+       8 covered mutants were not detected
       6 errors were encountered
       0 syntax errors were encountered
      14 time outs were encountered
     787 mutants required more time than configured
```

Also here, seems like a very good improvement with code reduction!

**Comment by Ocramius** [comment]
Thanks @staabm!

**Comment by Ocramius** [comment]
Btw: notice the reduction in mutations: that really says a lot whenever we refactor :)

**Pull request merged**